lowcar/devices/KoalaBear/KoalaBear.cpp

```diff
@@ -5,358 +5,346 @@
+    pinMode(AIN2, OUTPUT);
+    pinMode(BIN1, OUTPUT);
+    pinMode(BIN2, OUTPUT);
 }
 
 void KoalaBear::device_disable () {
-	// digitalWrite(SLEEP, HIGH);
-	//this->pid->setCoefficients(1, 0, 0);
-	//this->pid->resetEncoder();
-
-	this->desired_speeds[MTRA] = 0.0;
-	this->desired_speeds[MTRB] = 0.0;
-	this->pid_enabled[MTRA] = FALSE;
-	this->pid_enabled[MTRB] = FALSE;
+    // digitalWrite(SLEEP, HIGH);
+    //this->pid->setCoefficients(1, 0, 0);
+    //this->pid->resetEncoder();
+    
+    this->pid_a->set_coefficients(0.0, 0.0, 0.0);
+    this->pid_b->set_coefficients(0.0, 0.0, 0.0);
+    
+    this->target_speed_a = 0.0;
+    this->target_speed_b = 0.0;
+    this->pid_enabled_a = FALSE;
+    this->pid_enabled_b = FALSE;
 }
 
 void KoalaBear::device_actions () {
-	float target_A, target_B;
-
-	//switch between displaying info about MTRA and MTRB every 2 seconds
-	if (millis() - this->prev_led_time > 2000) {
-		this->curr_led_mtr = (this->curr_led_mtr == MTRA) ? MTRB : MTRA;
-		this->prev_led_time = millis();
-	}
-	this->led->ctrl_LEDs(this->desired_speeds[this->curr_led_mtr], TRUE); // TODO: Consider changing second arg to a condition
-
-	if (this->pid_enabled[MTRA]) {
-		//this->pid->setSetpoint(this->desired_speeds[MTRA]);
-		//target = this->pid->compute();
-		target_A = 0.0;
-	} else {
-		target_A = this->desired_speeds[MTRA];
-	}
-
-	if (this->pid_enabled[MTRB]) {
-		//this->pid->setSetpoint(this->desired_speeds[MTRB]);
-		//target = this->pid->compute();
-		target_B = 0.0;
-	} else {
-		target_B = this->desired_speeds[MTRB];
-	}
-
-	digitalWrite(SLEEP, HIGH);
-	digitalWrite(RESET, LOW);
-
-	//send target duty cycle to drive function
-	drive(target_A, MTRA);
-	drive(target_B *-1.0, MTRB); //MTRB by default moves in the opposite direction as motor A
+    float target_A, target_B;
+    unsigned long curr_time = millis();
+    
+    // switch between displaying info about MTRA and MTRB every 2 seconds
+    if (curr_time - this->prev_led_time > 2000) {
+        this->curr_led_mtr = (this->curr_led_mtr == MTRA) ? MTRB : MTRA;
+        this->prev_led_time = curr_time;
+    }
+    if (this->curr_led_mtr == MTRA) {
+        this->led->ctrl_LEDs(this->target_speed_a, this->deadband_a, TRUE);
+    } else {
+        this->led->ctrl_LEDs(this->target_speed_b, this->deadband_b, TRUE);
+    }
+    
+    if (this->pid_enabled_a) {
+        this->pid_a->set_target_speed(this->target_speed_a);
+        target_A = this->pid_a->compute((float) enc_a);
+    } else {
+        target_A = this->target_speed_a;
+    }
+
+    if (this->pid_enabled_b) {
+        this->pid_b->set_target_speed(this->target_speed_b);
+        target_B = this->pid_b->compute((float) enc_b);
+    } else {
+        target_B = this->target_speed_b;
+    }
+
+    digitalWrite(SLEEP, HIGH);
+    digitalWrite(RESET, LOW);
+
+    // send target duty cycle to drive function
+    drive(target_A * -1.0, MTRA); // by default, MTRA drives the wrong way
+    drive(target_B, MTRB);
```

Comment:
The motor directions should be device parameters

Reply:
1.0 feature? Since this would require an additional parameter in `runtime_util.c`?

Let's stick with this for now, I'll add this as an issue though that I can add when working with the encoders (I'm pretty sure that the PID math doesn't take this into account as of now either)

Reply:
I mean, KoalaBear needs a revamp anyways, so this can be part of that issue #92